### PR TITLE
Issue 16168 - isCopyable trait for value types

### PR DESCRIPTION
**Pull request by SealabJaster**
This is the first time I've contributed code to a project(other than a slight documentation improvement), so I'm sorry if I've messed up something.

https://issues.dlang.org/show_bug.cgi?id=16168

### CODE REVIEW

**Comment by wilzbach** [comment]
> This is the first time I've contributed code to a project(other than a slight documentation improvement)

That's great to hear!

> so I'm sorry if I've messed up something.

Do you see the failing CI integrations?
If you click on them you see why they are failing. It looks like it's just a [style](http://dlang.org/dstyle.html) issue ;-)

File: std/traits.d [hunk 1]
```diff
@@ -7029,3 +7030,41 @@ template isFinal(X...) if (X.length == 1)
...
+ + Copying can be disabled by using `@disable this(this)` for structs.
+ + 
+ + Params:
+ +  S = The value type (struct, int, etc.) to check.
+ +
+ + Returns:
+ +  `true` if `S` can be copied.
+ + ++/
+template isCopyable(S)
+if(!is(S == class))
```
Comment (MetaLang) [inline std/traits.d]: You should also check whether S is an interface using `!is(S == interface)`.

File: std/traits.d [hunk 2]
```diff
@@ -7029,3 +7030,41 @@ template isFinal(X...) if (X.length == 1)
...
+ + If a type cannot be copied, then code such as `MyStruct x; auto y = x;` will fail to compile.
+ + Copying can be disabled by using `@disable this(this)` for structs.
+ +
+ + Params:
+ +  S = The value type (struct, int, etc.) to check.
+ +
+ + Returns:
+ +  `true` if `S` can be copied.
+ + ++/
+template isCopyable(S)
```
Comment (andralex) [inline std/traits.d]:
```d
enum isCopyable(S) = is(typeof(
	{ S original; S copy = original; }
));
```

No need to restrict for classes and interfaces - references to those are copyable.

**Comment by andralex** [comment]
preapproved pending changes

**Comment by SealabJaster** [comment]
I'm not sure if it's worth mentioning, but using `S foo; S bar = foo;` causes `isCopyable` to be `false` when using `isCopyable!S2` in the unittest. But using `S foo = S.init; S bar = foo;` makes it `true`.

Example:
https://dpaste.dzfl.pl/10aa9f47979d

**Comment by japplegame** [comment]
`isCopyable` needs frame pointer, because S2 is nested. Just add `static` before struct definition and tests will pass.
Anyway you should not use `S foo;` or `S foo = S.init;` because the default constructor can be disabled and `S.init` can be overridden. Use `S foo = void;` instead.
Personally, I prefer
```d
enum isCopyConstructible(T, From = T) = is(typeof({ From from = void; T t = from; }));
```

**Pull request merged**